Can't build on GCC 14

Hi!

I was checking out libpostal, and saw something that could be improved.

---
#### My country is USA

---
#### Here's how I'm using libpostal
I'm trying to canonicalize addresses.

---
#### Here's what I did
`make`

---
#### Here's what I got
```
$ make
make[2]: Entering directory '/home/user/projects/libpostal/libpostal/src'
/bin/sh ../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I.. -I/usr/local/include    -Wall -Wextra -Wno-unused-function -Wformat -Werror=format-security -Winit-self -Wno-sign-compare -DLIBPOSTAL_DATA_DIR='"/home/user/projects/libpostal/libpostal_data/libpostal"' -g -mfpmath=sse -msse2 -DUSE_SSE -g -O2 -O2 -D LIBPOSTAL_EXPORTS  -MT libpostal_la-libpostal.lo -MD -MP -MF .deps/libpostal_la-libpostal.Tpo -c -o libpostal_la-libpostal.lo `test -f 'libpostal.c' || echo './'`libpostal.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I.. -I/usr/local/include -Wall -Wextra -Wno-unused-function -Wformat -Werror=format-security -Winit-self -Wno-sign-compare -DLIBPOSTAL_DATA_DIR=\"/home/user/projects/libpostal/libpostal_data/libpostal\" -g -mfpmath=sse -msse2 -DUSE_SSE -g -O2 -O2 -D LIBPOSTAL_EXPORTS -MT libpostal_la-libpostal.lo -MD -MP -MF .deps/libpostal_la-libpostal.Tpo -c libpostal.c  -fPIC -DPIC -o .libs/libpostal_la-libpostal.o
libpostal.c: In function ‘libpostal_classify_language’:
libpostal.c:217:58: error: initialization of ‘libpostal_language_classifier_response_t *’ {aka ‘struct libpostal_language_classifier_response *’} from incompatible pointer type ‘language_classifier_response_t *’ {aka ‘struct language_classifier_response *’} [-Wincompatible-pointer-types]
  217 |     libpostal_language_classifier_response_t *response = classify_languages(address);
      |                                                          ^~~~~~~~~~~~~~~~~~
make[2]: *** [Makefile:1417: libpostal_la-libpostal.lo] Error 1
make[2]: Leaving directory '/home/user/projects/libpostal/libpostal/src'
make[1]: *** [Makefile:470: all-recursive] Error 1
make[1]: Leaving directory '/home/user/projects/libpostal/libpostal'
make: *** [Makefile:379: all] Error 2


$ gcc --version
gcc (GCC) 14.2.1 20240910
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

---
#### Here's what I was expecting
The package to build successfully.

---
#### For parsing issues, please answer "yes" or "no" to all that apply.

Does not apply.

---
#### Here's what I think could be improved

>GCC no longer allows implicitly casting all pointer types to all other pointer types. This behavior is now restricted to the void * type and its qualified variations. 

https://gcc.gnu.org/gcc-14/porting_to.html

Instead, libpostal should probably just export the other type.

@lachesis, @mikicz  apart from the `incompatible pointer type` error, I also gets `implicit declaration of function` error for `strndup`

```
expand.c: In function 'expand_alternative_phrase_option':
expand.c:1502:36: error: implicit declaration of function 'strndup' [-Wimplicit-function-declaration]
 1502 |                 char *dupe_token = strndup(token + left_spaces, token_len - left_spaces - right_spaces);

numex.c: In function 'is_valid_roman_numeral':
numex.c:1216:18: error: implicit declaration of function 'strndup' [-Wimplicit-function-declaration]
 1216 |     char *copy = strndup(str, len);

```
To fix this issue, I manually #included `strndup.h` in those two files.

Did you get similar error while compiling the library?

My environment is MSYS2-MinGW64 with GCC 14.2.0

I was not able to build with that branch, with the same warning, but in a different place:

```bash
git clone https://github.com/openvenues/libpostal
git checkout lang_classifier_gcc14
./bootstrap.sh
./configure
make
```

```
gcc -DHAVE_CONFIG_H -I.. -I/usr/local/include    -Wall -Wextra -Wno-unused-function -Wformat -Werror=format-security -Winit-self -Wno-sign-compare -DLIBPOSTAL_DATA_DIR='"/usr/local/share/libpostal"' -g -mfpmath=sse -msse2 -DUSE_SSE -g -O2 -O3  -MT language_classifier_train-sparse_matrix_utils.o -MD -MP -MF .deps/language_classifier_train-sparse_matrix_utils.Tpo -c -o language_classifier_train-sparse_matrix_utils.o `test -f 'sparse_matrix_utils.c' || echo './'`sparse_matrix_utils.c
(... snip unrelated warning ...)
sparse_matrix_utils.c: In function ‘sparse_matrix_unique_columns’:
sparse_matrix_utils.c:100:50: error: passing argument 2 of ‘sparse_matrix_add_unique_columns’ from incompatible pointer type [-Wincompatible-pointer-types]
  100 |     if (sparse_matrix_add_unique_columns(matrix, unique_columns, ret)) {
      |                                                  ^~~~~~~~~~~~~~
      |                                                  |
      |                                                  kh_int_set_t *
```

However I think that's the only other place where there's an error. I have the following "patch" in my local branch:

```
diff --git a/src/libpostal.c b/src/libpostal.c
index 066a3015..4bdaeefd 100644
--- a/src/libpostal.c
+++ b/src/libpostal.c
@@ -214,7 +214,7 @@ libpostal_fuzzy_duplicate_status_t libpostal_is_street_duplicate_fuzzy(size_t nu
 }

 libpostal_language_classifier_response_t *libpostal_classify_language(char *address) {
-    libpostal_language_classifier_response_t *response = classify_languages(address);
+    libpostal_language_classifier_response_t *response = (libpostal_language_classifier_response_t*)(void*)classify_languages(address);

     if (response == NULL) {
         log_error("Language classification returned NULL\n");
diff --git a/src/sparse_matrix_utils.c b/src/sparse_matrix_utils.c
index 53fcaf97..53f1c3d6 100644
--- a/src/sparse_matrix_utils.c
+++ b/src/sparse_matrix_utils.c
@@ -97,7 +97,7 @@ uint32_array *sparse_matrix_unique_columns(sparse_matrix_t *matrix) {
     khash_t(int_set) *unique_columns = kh_init(int_set);
     uint32_array *ret = uint32_array_new();

-    if (sparse_matrix_add_unique_columns(matrix, unique_columns, ret)) {
+    if (sparse_matrix_add_unique_columns(matrix, (void*)unique_columns, ret)) {
         kh_destroy(int_set, unique_columns);
         return ret;
     }
```

My apologies for not calling out that location in the original ticket as well.

--- a/src/expand.c
+++ b/src/expand.c
@@ -1567,7 +1567,7 @@ cstring_array *expand_address_phrase_option(char *input, libpostal_normalize_opt
 
     size_t len = strlen(input);
 
-    language_classifier_response_t *lang_response = NULL;
+    libpostal_language_classifier_response_t *lang_response = NULL;
 
     if (options.num_languages == 0) {
          lang_response = classify_languages(input);
@@ -1627,7 +1627,7 @@ cstring_array *expand_address_phrase_option(char *input, libpostal_normalize_opt
     kh_destroy(str_set, unique_strings);
 
     if (lang_response != NULL) {
-        language_classifier_response_destroy(lang_response);
+        libpostal_language_classifier_response_destroy(lang_response);
     }
 
     char_array_destroy(temp_string);

--- a/src/language_classifier.c
+++ b/src/language_classifier.c
@@ -46,7 +46,7 @@ language_classifier_t *get_language_classifier(void) {
     return language_classifier;
 }
 
-void language_classifier_response_destroy(language_classifier_response_t *self) {
+void language_classifier_response_destroy(libpostal_language_classifier_response_t *self) {
     if (self == NULL) return;
     if (self->languages != NULL) {
         free(self->languages);
@@ -59,7 +59,7 @@ void language_classifier_response_destroy(language_classifier_response_t *self) 
     free(self);
 }
 
-language_classifier_response_t *classify_languages(char *address) {
+libpostal_language_classifier_response_t *classify_languages(char *address) {
     language_classifier_t *classifier = get_language_classifier();
     
     if (classifier == NULL) {
@@ -88,7 +88,7 @@ language_classifier_response_t *classify_languages(char *address) {
     size_t n = classifier->num_labels;
     double_matrix_t *p_y = double_matrix_new_zeros(1, n);
 
-    language_classifier_response_t *response = NULL;
+    libpostal_language_classifier_response_t *response = NULL;
     bool model_exp = false;
     if (classifier->weights_type == MATRIX_DENSE) {
         model_exp = logistic_regression_model_expectation(classifier->weights.dense, x, p_y);
@@ -129,7 +129,7 @@ language_classifier_response_t *classify_languages(char *address) {
 
         free(indices);
 
-        response = malloc(sizeof(language_classifier_response_t));
+        response = malloc(sizeof(libpostal_language_classifier_response_t));
         response->num_languages = num_languages;
         response->languages = languages;
         response->probs = probs;

--- a/src/language_classifier.h
+++ b/src/language_classifier.h
@@ -6,6 +6,8 @@
 #include <stdint.h>
 #include <stdbool.h>
 
+#include "libpostal.h"
+
 #include "collections.h"
 #include "language_features.h"
 #include "logistic_regression.h"
@@ -29,21 +31,14 @@ typedef struct language_classifier {
     } weights;
 } language_classifier_t;
 
-
-typedef struct language_classifier_response {
-    size_t num_languages;
-    char **languages;
-    double *probs;
-} language_classifier_response_t;
-
 // General usage
 
 language_classifier_t *language_classifier_new(void);
 language_classifier_t *get_language_classifier(void);
 language_classifier_t *get_language_classifier_country(void);
 
-language_classifier_response_t *classify_languages(char *address);
-void language_classifier_response_destroy(language_classifier_response_t *self);
+libpostal_language_classifier_response_t *classify_languages(char *address);
+void language_classifier_response_destroy(libpostal_language_classifier_response_t *self);
 
 void language_classifier_destroy(language_classifier_t *self);
 

--- a/src/language_classifier_cli.c
+++ b/src/language_classifier_cli.c
@@ -29,7 +29,7 @@ int main(int argc, char **argv) {
     }
 
 
-    language_classifier_response_t *response = classify_languages(address);
+    libpostal_language_classifier_response_t *response = classify_languages(address);
     if (response == NULL) {
         printf("Could not classify language\n");
         exit(EXIT_FAILURE);

--- a/src/language_classifier_test.c
+++ b/src/language_classifier_test.c
@@ -34,7 +34,7 @@ double test_accuracy(char *filename) {
             continue;
         }
 
-        language_classifier_response_t *response = classify_languages(address);
+        libpostal_language_classifier_response_t *response = classify_languages(address);
         if (response == NULL || response->num_languages == 0) {
             printf("%s\tNULL\t%s\n", language, address);
             continue;

--- a/src/libpostal.c
+++ b/src/libpostal.c
@@ -119,7 +119,7 @@ char **libpostal_near_dupe_hashes_languages(size_t num_components, char **labels
 
 
 char **libpostal_place_languages(size_t num_components, char **labels, char **values, size_t *num_languages) {
-    language_classifier_response_t *lang_response = place_languages(num_components, labels, values);
+    libpostal_language_classifier_response_t *lang_response = place_languages(num_components, labels, values);
     if (lang_response == NULL) {
         *num_languages = 0;
         return NULL;

--- a/src/near_dupe.c
+++ b/src/near_dupe.c
@@ -670,7 +670,7 @@ cstring_array *near_dupe_hashes_languages(size_t num_components, char **labels, 
 
     libpostal_normalize_options_t normalize_options = libpostal_get_default_options();
 
-    language_classifier_response_t *lang_response = NULL;
+    libpostal_language_classifier_response_t *lang_response = NULL;
 
     if (num_languages == 0) {
         lang_response = place_languages(num_components, labels, values);

--- a/src/place.c
+++ b/src/place.c
@@ -17,10 +17,10 @@ static inline bool is_address_text_component(char *label) {
             );
 }
 
-language_classifier_response_t *place_languages(size_t num_components, char **labels, char **values) {
+libpostal_language_classifier_response_t *place_languages(size_t num_components, char **labels, char **values) {
     if (num_components == 0 || values == NULL || labels == NULL) return NULL;
 
-    language_classifier_response_t *lang_response = NULL;
+    libpostal_language_classifier_response_t *lang_response = NULL;
     
     char *label;
     char *value;

--- a/src/place.h
+++ b/src/place.h
@@ -32,7 +32,7 @@ typedef struct place {
     char *website;
 } place_t;
 
-language_classifier_response_t *place_languages(size_t num_components, char **labels, char **values);
+libpostal_language_classifier_response_t *place_languages(size_t num_components, char **labels, char **values);
 
 place_t *place_new(void);
 

--- a/src/sparse_matrix_utils.c
+++ b/src/sparse_matrix_utils.c
@@ -94,15 +94,15 @@ inline bool sparse_matrix_add_unique_columns_alias(sparse_matrix_t *matrix, khas
 }
 
 uint32_array *sparse_matrix_unique_columns(sparse_matrix_t *matrix) {
-    khash_t(int_set) *unique_columns = kh_init(int_set);
+    khash_t(int_uint32) *unique_columns = kh_init(int_uint32);
     uint32_array *ret = uint32_array_new();
 
     if (sparse_matrix_add_unique_columns(matrix, unique_columns, ret)) {
-        kh_destroy(int_set, unique_columns);
+        kh_destroy(int_uint32, unique_columns);
         return ret;
     }
 
-    kh_destroy(int_set, unique_columns);
+    kh_destroy(int_uint32, unique_columns);
     uint32_array_destroy(ret);
     return NULL;
 }

